core/src/main/scala/kafka/log/Log.scala

```diff
@@ -583,6 +578,58 @@ class Log(val dir: File,
   }
 
   /**
+    * Find any segments that match the user-supplied predicate UNLESS it is the final segment
+    * and it is empty (since we would just end up re-creating it)
+    * @param predicate A function that takes in a single log segment and returns true iff it is deletable
+    * @return the segments ready to be deleted
+    */
+  private def deletableSegments(predicate: LogSegment => Boolean) = {
+    val lastEntry = segments.lastEntry
+    if (lastEntry == null) Seq.empty
+    else logSegments.takeWhile(s => predicate(s) && (s.baseOffset != lastEntry.getValue.baseOffset || s.size > 0))
+  }
+
+  /**
+    * Delete any log segments that have either expired due to time based retention
+    * or because the log size is > retentionSize
+    */
+  def deleteOldSegments(): Int = {
+    if (!config.delete) return 0
+    retentionMsExpiredSegments(deleteOldSegments) + retentionSizeBreachedSegments(deleteOldSegments)
+  }
+
+  private def hasOldSegments(predicate: LogSegment => Boolean): Int = {
+    lock synchronized {
+      deletableSegments(predicate).size
+    }
+  }
+
+  private def retentionMsExpiredSegments(runWithPredicate: (LogSegment => Boolean) => Int) : Int = {
+    if (config.retentionMs < 0) return 0
+    val startMs = time.milliseconds
+    runWithPredicate(startMs - _.lastModified > config.retentionMs)
+  }
+
+  private def retentionSizeBreachedSegments(runWithPredicate: (LogSegment => Boolean) => Int) : Int = {
+    if (config.retentionSize < 0 || size < config.retentionSize) return 0
+    var diff = size - config.retentionSize
+    def shouldDelete(segment: LogSegment) = {
+      if (diff - segment.size >= 0) {
+        diff -= segment.size
+        true
+      } else {
+        false
+      }
+    }
+    runWithPredicate(shouldDelete)
+  }
+
+  def shouldDeleteSegments: Boolean = {
+    if (!config.delete) return false
+    retentionMsExpiredSegments(hasOldSegments) + retentionSizeBreachedSegments(hasOldSegments) > 0
```

Comment:
Could we just pass in deletableSegments instead of hasOldSegments and then get rid of hasOldSegments?